packages/vite/src/node/server/moduleGraph.ts

```diff
@@ -36,6 +36,28 @@ export class ModuleNode {
   ssrError: Error | null = null
   lastHMRTimestamp = 0
   lastInvalidationTimestamp = 0
+  /**
+   * If the module only needs to update its imports timestamp (e.g. within an HMR chain),
+   * it is considered soft-invalidated. In this state, its `transformResult` should exist,
+   * and the next `transformRequest` for this module will replace the timestamps.
+   *
+   * By default the value is `undefined` if it's not soft/hard-invalidated. If it gets
+   * soft-invalidated, this will contain the previous `transformResult` value. If it gets
+   * hard-invalidated, this will be set to `null`.
+   * @internal
+   */
+  softInvalidatedTransformResult?: TransformResult | null
```

Comment:
I'm fine with the three state field but if doesn't requires to many changes I like to use:
`softInvalidatedTransformResult: TransformResult | null | undefined`

Other solution is to use a string constant instead of the null marker:
`softInvalidatedTransformResult: TransformResult | "HARD_INVALIDATED" | undefined`

Reply:
I like the explicitness. I'll try the latter suggestion since that should make it easier to read.